Add INCLUDING GENERATED in includes for postgresql 12

Dishwasha opened the pull request:
When using generated columns now available in postgresql 12 https://www.postgresql.org/docs/12/ddl-generated-columns.html, pg_partman creates new tables and partitions without copying the generated expressions.  This adds the "INCLUDING GENERATED" like_option only if the postgres version is at least 12.

Dishwasha commented:
I tried following the testing README to use pgTap to validate these changes, but I wasn't able to figure out how to get it to successfully run the tests.

keithf4 commented:
Thanks for reporting this! Adapted your changes into an extension update file in the following gist link. Mind giving this a test to see if it all works ok?

https://gist.github.com/keithf4/f36693a1a8eacc41fd11f34746faf756

Ensure file is named: `pg_partman--4.3.0--4.3.1.sql` and is placed in your `$SHAREDIR/extension` folder so that the ALTER EXTENSION command can pick it up. You can see your $SHAREDIR value by running `pg_config`. Note this will require that 4.3.0 is installed already.

Dishwasha commented:
I'm not sure if I am doing something wrong.

- I removed my custom patch and did an apt reinstall of the pg_partman extension just to be safe
- I put that gist at /usr/share/postgresql/12/extension/pg_partman--4.3.0--4.3.1.sql
- I full stopped and started postgresql
- Then did ` ALTER EXTENSION pg_partman UPDATE TO '4.3.1'` from postgres
- dropped and recreated both the table and partman schema with extension.
- `SELECT partman.create_parent('<table>', 'time', 'native', 'weekly');`

Result: I don't see any generated expressions under defaults in the partitions.

I put my patch back in, downgraded the extension back to 4.3.0 and re-created everything and I see the generated columns in the partitions.

Am I supposed to have a pg_partman--4.3.1.sql as well?  Feel free to review my patch at https://gist.github.com/Dishwasha/50bf99da4d1e5e1b107fa93fce4a4586 which I put as pg_partman--4.3.0.sql and can be diffed with original.

keithf4 commented:
You have to have 4.3.0 installed already and then place the update file in your $SHAREDIR/extension location. You don't need a pg_partman--4-3.1.sql file yet since you're not installing 4.3.0 initially. Just doing an update from 4.3.0.

Note that IDENTITY columns are special types of GENERATED columns and right now I don't include them as being inherited. When used with natively partitioned tables, the IDENTITY existing on the parent will pass down values to the child tables. So if I read correctly, if IDENTITY was applied to each child as well, you'd have a unique sequence on each child table vs having a consistent one across the whole partition set. They're actually a completely separate INCLUDING clause to LIKE, probably for something similar to this reason. This does mean that if you use IDENTITY you **must** do all writes through the parent of the partition set.

I did test to see if a regular GENERATED column existing only on the parent passes down the generated values to its children like IDENTITY does. It does not.

So, as you likely found as well, a normal GENERATED column does need to be applied to each child. Here's an example with Fahrenheit  to Celsius conversion that seems to work with this update file. This also shows how IDENTITY works as well. I mostly just implemented your changes with some additional space padding added in around the string concatenation, so I hope it works:

Before update:
```
keith@keith=# create extension pg_partman schema partman;
CREATE EXTENSION
Time: 90.065 ms
keith@keith=# \dx 
                                   List of installed extensions
    Name    | Version |   Schema   |                         Description                          
------------+---------+------------+--------------------------------------------------------------
 dblink     | 1.2     | public     | connect to other PostgreSQL databases from within a database
 pg_partman | 4.3.0   | partman    | Extension to manage partitioned tables by time or ID
 pgtap      | 1.1.0   | public     | Unit testing for PostgreSQL
 plpgsql    | 1.0     | pg_catalog | PL/pgSQL procedural language
(4 rows)

```
After update
```
keith@keith=# ALTER EXTENSION pg_partman UPDATE TO '4.3.1';
ALTER EXTENSION
Time: 23.520 ms
keith@keith=# \dx
                                   List of installed extensions
    Name    | Version |   Schema   |                         Description                          
------------+---------+------------+--------------------------------------------------------------
 dblink     | 1.2     | public     | connect to other PostgreSQL databases from within a database
 pg_partman | 4.3.1   | partman    | Extension to manage partitioned tables by time or ID
 pgtap      | 1.1.0   | public     | Unit testing for PostgreSQL
 plpgsql    | 1.0     | pg_catalog | PL/pgSQL procedural language
(4 rows)
```

New table
```
keith@keith=# CREATE TABLE measurement (                                                       
    city_id         int GENERATED ALWAYS AS IDENTITY,
    logdate         date not null default now(),
    peaktemp_f        int,
    peaktemp_c  int generated always as ((peaktemp_f - 32) * 5/9) stored
) PARTITION BY RANGE (logdate)
;
CREATE TABLE
Time: 13.710 ms
```
Create partitioned table and test
```
keith@keith=# select partman.create_parent('public.measurement', 'logdate', 'native', 'daily');
 create_parent 
---------------
 t
(1 row)

Time: 66.603 ms
keith@keith=# \d+ measurement;
                                                   Partitioned table "public.measurement"
   Column   |  Type   | Collation | Nullable |                        Default                         | Storage | Stats target | Description
 
------------+---------+-----------+----------+--------------------------------------------------------+---------+--------------+------------
-
 city_id    | integer |           | not null | generated always as identity                           | plain   |              | 
 logdate    | date    |           | not null | now()                                                  | plain   |              | 
 peaktemp_f | integer |           |          |                                                        | plain   |              | 
 peaktemp_c | integer |           |          | generated always as ((peaktemp_f - 32) * 5 / 9) stored | plain   |              | 
Partition key: RANGE (logdate)
Partitions: measurement_p2020_04_05 FOR VALUES FROM ('2020-04-05') TO ('2020-04-06'),
            measurement_p2020_04_06 FOR VALUES FROM ('2020-04-06') TO ('2020-04-07'),
            measurement_p2020_04_07 FOR VALUES FROM ('2020-04-07') TO ('2020-04-08'),
            measurement_p2020_04_08 FOR VALUES FROM ('2020-04-08') TO ('2020-04-09'),
            measurement_p2020_04_09 FOR VALUES FROM ('2020-04-09') TO ('2020-04-10'),
            measurement_p2020_04_10 FOR VALUES FROM ('2020-04-10') TO ('2020-04-11'),
            measurement_p2020_04_11 FOR VALUES FROM ('2020-04-11') TO ('2020-04-12'),
            measurement_p2020_04_12 FOR VALUES FROM ('2020-04-12') TO ('2020-04-13'),
            measurement_p2020_04_13 FOR VALUES FROM ('2020-04-13') TO ('2020-04-14'),
            measurement_default DEFAULT

keith@keith=# \d+ measurement_p2020_04_05 
                                                   Table "public.measurement_p2020_04_05"
   Column   |  Type   | Collation | Nullable |                        Default                         | Storage | Stats target | Description
 
------------+---------+-----------+----------+--------------------------------------------------------+---------+--------------+------------
-
 city_id    | integer |           | not null |                                                        | plain   |              | 
 logdate    | date    |           | not null | now()                                                  | plain   |              | 
 peaktemp_f | integer |           |          |                                                        | plain   |              | 
 peaktemp_c | integer |           |          | generated always as ((peaktemp_f - 32) * 5 / 9) stored | plain   |              | 
Partition of: measurement FOR VALUES FROM ('2020-04-05') TO ('2020-04-06')
Partition constraint: ((logdate IS NOT NULL) AND (logdate >= '2020-04-05'::date) AND (logdate < '2020-04-06'::date))
Access method: heap

keith@keith=# \d+ measurement_default 
                                                     Table "public.measurement_default"
   Column   |  Type   | Collation | Nullable |                        Default                         | Storage | Stats target | Description
 
------------+---------+-----------+----------+--------------------------------------------------------+---------+--------------+------------
-
 city_id    | integer |           | not null |                                                        | plain   |              | 
 logdate    | date    |           | not null | now()                                                  | plain   |              | 
 peaktemp_f | integer |           |          |                                                        | plain   |              | 
 peaktemp_c | integer |           |          | generated always as ((peaktemp_f - 32) * 5 / 9) stored | plain   |              | 
Partition of: measurement DEFAULT
Partition constraint: (NOT ((logdate IS NOT NULL) AND (((logdate >= '2020-04-05'::date) AND (logdate < '2020-04-06'::date)) OR ((logdate >= 
'2020-04-06'::date) AND (logdate < '2020-04-07'::date)) OR ((logdate >= '2020-04-07'::date) AND (logdate < '2020-04-08'::date)) OR ((logdate
 >= '2020-04-08'::date) AND (logdate < '2020-04-09'::date)) OR ((logdate >= '2020-04-09'::date) AND (logdate < '2020-04-10'::date)) OR ((log
date >= '2020-04-10'::date) AND (logdate < '2020-04-11'::date)) OR ((logdate >= '2020-04-11'::date) AND (logdate < '2020-04-12'::date)) OR (
(logdate >= '2020-04-12'::date) AND (logdate < '2020-04-13'::date)) OR ((logdate >= '2020-04-13'::date) AND (logdate < '2020-04-14'::date)))
))
Access method: heap

keith@keith=# insert into measurement (peaktemp_f) values (32);
INSERT 0 1
Time: 13.908 ms
keith@keith=# select * from measurement;
 city_id |  logdate   | peaktemp_f | peaktemp_c 
---------+------------+------------+------------
       1 | 2020-04-09 |         32 |          0
(1 row)

Time: 11.462 ms
keith@keith=# insert into measurement (peaktemp_f) values (95);
INSERT 0 1
Time: 8.708 ms
keith@keith=# select * from measurement;
 city_id |  logdate   | peaktemp_f | peaktemp_c 
---------+------------+------------+------------
       1 | 2020-04-09 |         32 |          0
       2 | 2020-04-09 |         95 |         35
(2 rows)

Time: 6.971 ms
```

Dishwasha commented:
Okay, I just re-did my steps and now it's showing generated columns on my partitions as expected.  I suspect I neglected to re-run alter extension after dropping and re-creating the partman schema.  My apologies for misreporting on my last comment and thank you for getting this included.

keithf4 commented:
Glad to hear it worked! Will try and get a new version out next week.

keithf4 commented:
Version 4.3.1 has been released and includes this fix. Thanks for reporting it!
